DMVM-137 feat: 인증/인가

emibgo2 opened the pull request:
## 🎫 137

- ✨ 새로운 기능 추가

## 변경 사항에 대한 설명
![image](https://github.com/PBTP/mongle-server/assets/73595178/719378cd-debd-4adf-a5d7-2de1eb4d8776)

**인증 기능을 추가하였습니다. (인가 X)**

accessToken은 Redis에서 관리하고
refreshToken은 DB에서 관리하도록 하였습니다.
이때문에 customer entity에 refresh_token 컬럼이 추가되었습니다. (text, not null, unique)

uuid는 OAuth Resource Server에서 제공해주는 identifier 값을 넣도록 하였습니다.
현재 제공하는 서비스는 Apple, Kakao 입니다. 
해당 두개의 서비스의 identifier값의 길이가 달라 char(20)에서 varchar(20)으로 변경하였습니다.


JWT Rotate 방식으로 하여금  refreshToken 또한 accessToken 재발급시에 재생성 되도록 하였습니다.


## 테스트 방법

본인의 정보를 입력합니다.
1. DB에 사용자가 있으면 토큰을 발급
2. DB에 사용자가 없으면 사용자 생성 후 토큰 발급
3. access_token 만료시 refresh_token으로 재발급
4. 한번 사용한 refresh_token은 사용 되지 않음 (401, Unauthorized)

## 변경된 환경

기존의 DB 정보를 datasource로 설정되었었는데 Redis 정보 파라미터와 일관성을 맞추기 위해 datasource/db 로 변경되었습니다.

## 참고 사항

Elasticache 접속 방법도 DB 접속 방법과 동일합니다.

emibgo2 committed:
ㄹㄷㅁㅅ

shine-jung commented on package.json:
```diff
@@ -20,11 +20,25 @@
+    "@nestjs/typeorm": "^10.0.2",
+    "class-transformer": "^0.5.1",
+    "class-validator": "^0.14.1",
+    "npm": "^10.8.0",
```
저희 npm도 yarn과 같이 사용하나요?

emibgo2 replied:
실수로 들어간것 같네요.. 뺐습니다!

shine-jung commented on package.json:
```diff
@@ -20,11 +20,25 @@
+    "class-validator": "^0.14.1",
+    "npm": "^10.8.0",
+    "passport": "^0.7.0",
+    "passport-jwt": "^4.0.1",
```
`passport-jwt` 사용 좋은 선택인 것 같습니다.

emibgo2 committed:
DMVM-137 feat: access token 발급

Commit:
feat: AccessToken, RefreshToken 검증 로직 추가

Commit:
feat: AccessToken 재발급 로직 추가

Commit:
Merge branch 'dev' of https://github.com/PBTP/mongle-server into feat-authorization

Commit:
refactor: CustomerDto validation annotation 추가
refactor: ValidationPipe global pipe로 추가

Commit:
rename: auth, customer domain 폴더 구조 변경

Commit:
Merge branch 'refs/heads/dev' into feat-authorazation

# Conflicts:
#	src/env/ssm-config.service.ts

Commit:
feat: 본인 정보 조회 API 추가
feat: ClassSerializerInterceptor, GlobalInterceptor로 추가
fix: loadParameterStoreValue 수정
- parameter store path로 조회시 가져올 수 있는 수가 제한(최대 10개)이 있어서 모든 파라미터를 가져올 수 있도록 수정
fix: jwt관련 환경변수 경로 변경

emibgo2 committed:
Merge branch 'dev' into feat-authorazation

Commit:
Refactor: accessToken 유니크하게 처리

Commit:
Refactor: accessToken 유니크하게 처리

shine-jung reviewed (approved):
고생하셨습니다
entity snakeCase와 global validation pipe 처리도 해주셨네요
추후 spec 파일 보강하는 것으로 하겠습니다!
